diff --git a/SignalsmithStretch.js b/SignalsmithStretch.js
index 7b39b2ae500c8563500c20a4f951878e309d845f..148614b7f3761d8b87cd0795723c47f431051f2e 100644
--- a/SignalsmithStretch.js
+++ b/SignalsmithStretch.js
@@ -84,17 +84,17 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					return result;
 				},
 				schedule: (objIn, adjustPrevious) => {
-					let outputTime = ('outputTime' in objIn) ? objIn.outputTime : currentTime;
+					let scheduledOutputTime = ('outputTime' in objIn) ? objIn.outputTime : (('output' in objIn) ? objIn.output : currentTime);
 
 					let latestSegment = this.timeMap[this.timeMap.length - 1];
-					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= outputTime) {
+					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= scheduledOutputTime) {
 						latestSegment = this.timeMap.pop();
 					}
 
 					let obj = Object.assign({}, latestSegment);
 					Object.assign(obj, {
 						input: null,
-						output: outputTime,
+						output: scheduledOutputTime,
 					});
 					Object.assign(obj, objIn);
 					if (obj.input === null) {
@@ -114,12 +114,12 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					}
 	
 					let currentMapSegment = this.timeMap[0];
-					while (this.timeMap.length > 1 && this.timeMap[1].output <= outputTime) {
+					while (this.timeMap.length > 1 && this.timeMap[1].output <= currentTime) {
 						this.timeMap.shift();
 						currentMapSegment = this.timeMap[0];
 					}
 					let rate = (currentMapSegment.active ? currentMapSegment.rate : 0);
-					let inputTime = currentMapSegment.input + (outputTime - currentMapSegment.output)*rate;
+					let inputTime = currentMapSegment.input + (currentTime - currentMapSegment.output)*rate;
 					this.timeIntervalCounter = this.timeIntervalSamples;
 					this.port.postMessage(['time', inputTime]);
 					
diff --git a/SignalsmithStretch.mjs b/SignalsmithStretch.mjs
index 7d5cae72e77a84143ed91ba12fc91b806dc4d9d1..fb421976aca16729ffa9ba22f8cd477c673c9b36 100644
--- a/SignalsmithStretch.mjs
+++ b/SignalsmithStretch.mjs
@@ -85,17 +85,17 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					return result;
 				},
 				schedule: (objIn, adjustPrevious) => {
-					let outputTime = ('outputTime' in objIn) ? objIn.outputTime : currentTime;
+					let scheduledOutputTime = ('outputTime' in objIn) ? objIn.outputTime : (('output' in objIn) ? objIn.output : currentTime);
 
 					let latestSegment = this.timeMap[this.timeMap.length - 1];
-					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= outputTime) {
+					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= scheduledOutputTime) {
 						latestSegment = this.timeMap.pop();
 					}
 
 					let obj = Object.assign({}, latestSegment);
 					Object.assign(obj, {
 						input: null,
-						output: outputTime,
+						output: scheduledOutputTime,
 					});
 					Object.assign(obj, objIn);
 					if (obj.input === null) {
@@ -115,12 +115,12 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					}
 	
 					let currentMapSegment = this.timeMap[0];
-					while (this.timeMap.length > 1 && this.timeMap[1].output <= outputTime) {
+					while (this.timeMap.length > 1 && this.timeMap[1].output <= currentTime) {
 						this.timeMap.shift();
 						currentMapSegment = this.timeMap[0];
 					}
 					let rate = (currentMapSegment.active ? currentMapSegment.rate : 0);
-					let inputTime = currentMapSegment.input + (outputTime - currentMapSegment.output)*rate;
+					let inputTime = currentMapSegment.input + (currentTime - currentMapSegment.output)*rate;
 					this.timeIntervalCounter = this.timeIntervalSamples;
 					this.port.postMessage(['time', inputTime]);
 					
