diff --git a/SignalsmithStretch.js b/SignalsmithStretch.js
index 7b39b2ae500c8563500c20a4f951878e309d845f..33712aac9325b10074ee8dd7342286e3874567f3 100644
--- a/SignalsmithStretch.js
+++ b/SignalsmithStretch.js
@@ -23,6 +23,7 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 			super(options);
 			this.wasmReady = false;
 			this.wasmModule = null;
+			this.disposed = false;
 			this.channels = 0;
 			this.buffersIn = [];
 			this.buffersOut = [];
@@ -84,17 +85,17 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					return result;
 				},
 				schedule: (objIn, adjustPrevious) => {
-					let outputTime = ('outputTime' in objIn) ? objIn.outputTime : currentTime;
+					let scheduledOutputTime = ('outputTime' in objIn) ? objIn.outputTime : (('output' in objIn) ? objIn.output : currentTime);
 
 					let latestSegment = this.timeMap[this.timeMap.length - 1];
-					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= outputTime) {
+					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= scheduledOutputTime) {
 						latestSegment = this.timeMap.pop();
 					}
 
 					let obj = Object.assign({}, latestSegment);
 					Object.assign(obj, {
 						input: null,
-						output: outputTime,
+						output: scheduledOutputTime,
 					});
 					Object.assign(obj, objIn);
 					if (obj.input === null) {
@@ -114,12 +115,12 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					}
 	
 					let currentMapSegment = this.timeMap[0];
-					while (this.timeMap.length > 1 && this.timeMap[1].output <= outputTime) {
+					while (this.timeMap.length > 1 && this.timeMap[1].output <= currentTime) {
 						this.timeMap.shift();
 						currentMapSegment = this.timeMap[0];
 					}
 					let rate = (currentMapSegment.active ? currentMapSegment.rate : 0);
-					let inputTime = currentMapSegment.input + (outputTime - currentMapSegment.output)*rate;
+					let inputTime = currentMapSegment.input + (currentTime - currentMapSegment.output)*rate;
 					this.timeIntervalCounter = this.timeIntervalSamples;
 					this.port.postMessage(['time', inputTime]);
 					
@@ -154,6 +155,16 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 						transfer: transfer
 					};
 				},
+				dispose: _ => {
+					let buffers = this.audioBuffers.flat(1).map(b => b.buffer);
+					this.audioBuffers = [];
+					this.audioBuffersStart = this.audioBuffersEnd = 0;
+					this.disposed = true;
+					return {
+						value: undefined,
+						transfer: buffers
+					};
+				},
 				addBuffers: sampleBuffers => {
 					sampleBuffers = [].concat(sampleBuffers);
 					this.audioBuffers.push(sampleBuffers);
@@ -232,6 +243,7 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 		}
 
 		process(inputList, outputList, parameters) {
+			if (this.disposed) return false;
 			if (!this.wasmReady) {
 				outputList.forEach(output => {
 					output.forEach(channel => {
diff --git a/SignalsmithStretch.mjs b/SignalsmithStretch.mjs
index 7d5cae72e77a84143ed91ba12fc91b806dc4d9d1..d73cd992ec37d6fa3cc6d8bc2e81755d490e3969 100644
--- a/SignalsmithStretch.mjs
+++ b/SignalsmithStretch.mjs
@@ -24,6 +24,7 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 			super(options);
 			this.wasmReady = false;
 			this.wasmModule = null;
+			this.disposed = false;
 			this.channels = 0;
 			this.buffersIn = [];
 			this.buffersOut = [];
@@ -85,17 +86,17 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					return result;
 				},
 				schedule: (objIn, adjustPrevious) => {
-					let outputTime = ('outputTime' in objIn) ? objIn.outputTime : currentTime;
+					let scheduledOutputTime = ('outputTime' in objIn) ? objIn.outputTime : (('output' in objIn) ? objIn.output : currentTime);
 
 					let latestSegment = this.timeMap[this.timeMap.length - 1];
-					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= outputTime) {
+					while (this.timeMap.length && this.timeMap[this.timeMap.length - 1].output >= scheduledOutputTime) {
 						latestSegment = this.timeMap.pop();
 					}
 
 					let obj = Object.assign({}, latestSegment);
 					Object.assign(obj, {
 						input: null,
-						output: outputTime,
+						output: scheduledOutputTime,
 					});
 					Object.assign(obj, objIn);
 					if (obj.input === null) {
@@ -115,12 +116,12 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 					}
 	
 					let currentMapSegment = this.timeMap[0];
-					while (this.timeMap.length > 1 && this.timeMap[1].output <= outputTime) {
+					while (this.timeMap.length > 1 && this.timeMap[1].output <= currentTime) {
 						this.timeMap.shift();
 						currentMapSegment = this.timeMap[0];
 					}
 					let rate = (currentMapSegment.active ? currentMapSegment.rate : 0);
-					let inputTime = currentMapSegment.input + (outputTime - currentMapSegment.output)*rate;
+					let inputTime = currentMapSegment.input + (currentTime - currentMapSegment.output)*rate;
 					this.timeIntervalCounter = this.timeIntervalSamples;
 					this.port.postMessage(['time', inputTime]);
 					
@@ -155,6 +156,16 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 						transfer: transfer
 					};
 				},
+				dispose: _ => {
+					let buffers = this.audioBuffers.flat(1).map(b => b.buffer);
+					this.audioBuffers = [];
+					this.audioBuffersStart = this.audioBuffersEnd = 0;
+					this.disposed = true;
+					return {
+						value: undefined,
+						transfer: buffers
+					};
+				},
 				addBuffers: sampleBuffers => {
 					sampleBuffers = [].concat(sampleBuffers);
 					this.audioBuffers.push(sampleBuffers);
@@ -233,6 +244,7 @@ function registerWorkletProcessor(Module, audioNodeKey) {
 		}
 
 		process(inputList, outputList, parameters) {
+			if (this.disposed) return false;
 			if (!this.wasmReady) {
 				outputList.forEach(output => {
 					output.forEach(channel => {
